packages/d3fc-webgl/src/series/glCandlestick.js

```diff
@@ -0,0 +1,252 @@
+import attributeBuilder from '../buffers/attributeBuilder';
+import glScaleBase from '../scale/glScaleBase';
+import programBuilder from '../program/programBuilder';
+import rectShader from '../shaders/rect/shader';
+import lineWidthShader from '../shaders/lineWidth';
+import drawModes from '../program/drawModes';
+import { rebind } from '@d3fc/d3fc-rebind';
+
+export default () => {
+    const program = programBuilder();
+    let xScale = glScaleBase();
+    let yScale = glScaleBase();
+    const lineWidth = lineWidthShader();
+    let decorate = () => {};
+
+    const xValueAttrib = 'aXValue';
+    const yValueAttrib = 'aYValue';
+    const xDirectionAttrib = 'aXDirection';
+    const yDirectionAttrib = 'aYDirection';
+    const bandwidthAttrib = 'aBandwidth';
+    const colorIndicatorAttrib = 'aColorIndicator';
+
+    const verticesPerElement = 12;
+
+    const draw = numElements => {
+        const shaderBuilder = rectShader();
+        program
+            .vertexShader(shaderBuilder.vertex())
+            .fragmentShader(shaderBuilder.fragment())
+            .mode(drawModes.TRIANGLES);
+
+        xScale.coordinate(0);
+        xScale(program);
+        yScale.coordinate(1);
+        yScale(program);
+
+        lineWidth(program);
+        setColors(numElements);
+        setYDirections(numElements);
+
+        program.vertexShader().appendBody(`
+        gl_Position.x += ((uLineWidth * aXDirection / 2.0) + (aBandwidth / 2.0)) / uScreen.x;
+        gl_Position.y += (uLineWidth * aYDirection / 2.0) / uScreen.y;
+      `);
+
+        decorate(program);
+
+        program(numElements * verticesPerElement);
+    };
+
+    draw.xValues = (...args) => {
+        const builder = program.buffers().attribute(xValueAttrib);
+        let xArray = new Float32Array(args[0].length * verticesPerElement);
+        xArray = xArray.map(
```

Comment:
Just having a quick poke around from a performance perspective - this method of mapping from the `args[0]` array to `xArray` is a bit slow (also I think it also replaces the array constructed in the line above). By my (crude) performance measurements, the following is approx x10 faster:

~~~
for (let i=0;i < xArray.length;i++) {
  xArray[i] = args[0][Math.floor(i / verticesPerElement)];
}
~~~

although YMMV.

Reply:
A similar map-based pattern is used elsewhere, so probably worth investigating further.

Reply:
Good spot. This should all hopefully be addressed by the changes I'm currently making to introduce 'projectedAttributeBuilder`.

Reply:
As this is an internal implementation detail of `d3fc-webgl`, are you ok with this change being made subsequently?